### PR TITLE
refactor: PluginManager Releasing Resources

### PR DESCRIPTION
**Pull request by ColinKYuen**
### Summary

PluginManager Release Resources

### Description

Added
- User method to call to release all resources held up by `ConnectionPluginManager`
- Added tracking of `ClusterAwareConnectionProxy` to `AbandonedConnectionCleanupThread` to be clear `ConnectionPluginManager`

### Additional Reviewers

@karenc-bq 
@sergiyv-bitquill

### CODE REVIEW

File: src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    @Test
+    void testReleasingAllWithEmptyList() {
+        ConnectionPluginManager.releaseAllResources();
+    }
+
+    @Test
+    void testReleasingAllWithEmptyListUsingDriver() {
+        Driver.releasePluginManager();
+    }
```
Comment (karenc-bq) [inline src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java:46]: Are these tests different?

File: src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+    @Test
+    void testReleasingAllWithInstanceInList() {
+        for (int i = 0; i < 5; i++) {
+            createPluginManager();
+        }
+
+        ConnectionPluginManager.releaseAllResources();
```
Comment (karenc-bq) [inline src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java:54]: What are the goals of these tests? We are not checking anything.

File: src/main/user-impl/java/com/mysql/cj/jdbc/AbandonedConnectionCleanupThread.java [hunk 3]
```diff
@@ -39,17 +45,16 @@
...
-import com.mysql.cj.protocol.NetworkResources;
-
 /**
  * This class implements a thread that is responsible for closing abandoned MySQL connections, i.e., connections that are not explicitly closed.
  * There is only one instance of this class and there is a single thread to do this task. This thread's executor is statically referenced in this same class.
  */
 public class AbandonedConnectionCleanupThread implements Runnable {
     private static final Set<ConnectionFinalizerPhantomReference> connectionFinalizerPhantomRefs = ConcurrentHashMap.newKeySet();
-    private static final ReferenceQueue<MysqlConnection> referenceQueue = new ReferenceQueue<>();
+    private static final ReferenceQueue<MysqlConnection> connectionReferenceQueue = new ReferenceQueue<>();
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/AbandonedConnectionCleanupThread.java:54]: I don't recommend adding these changes until we have tested it to know it does what we want it to do. I recommend putting this in a separate PR so we can merge the other changes in first.

File: src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import org.junit.jupiter.api.TestMethodOrder;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+import software.aws.rds.jdbc.mysql.Driver;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.spy;
+
+@TestMethodOrder(MethodOrderer.Alphanumeric.class)
```
Comment (karenc-bq) [inline src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java:20]: Why does these tests need to run in alphanumeric order?

File: src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    void setUp() {
+        openMocks = MockitoAnnotations.openMocks(this);
+        Mockito.doNothing()
+            .when(plugin).releaseResources();
+    }
+
+    @Test
+    void test_1_ReleasingAllWithEmptyList() {
+        ConnectionPluginManager.releaseAllResources();
+        Assert.assertEquals(0, ConnectionPluginManager.instances.size());
```
Comment (karenc-bq) [inline src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java:41]: 1. This is not really the correct assertEquals, IntelliJ tells you that too
image
2. Can we use static import
3. Still not sure if we need this test at all. `ConnectionPluginManager.instances.size()` is 0 to begin with in this test anyway. If I look at the implementation code of `releaseAllResources` it just has a for-each loop, if instances is empty it won't even enter the loop.  Not sure if this test adds any value, other than making sure Java for-each loop works.

File: src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    @Test
+    void test_2_ReleasingAllWithEmptyListUsingDriver() {
+        Driver.releasePluginManager();
+        Assert.assertEquals(0, ConnectionPluginManager.instances.size());
```
Comment (karenc-bq) [inline src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java:47]: Not sure we need all the `usingDriver` tests at all, `Driver.releasePluginManager` just calls `ConnectionPluginManager.releaseAllResources()`, which we are already testing.
Furthermore this is a `ConnectionPluginManagerTest`

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java [hunk 7]
```diff
@@ -47,18 +49,22 @@
   protected PropertySet propertySet = null;
   protected IConnectionPlugin headPlugin = null;
   ClusterAwareConnectionProxy proxy;
-  private final Thread shutdownHook = new Thread(this::releaseResources);
+  protected static final List<ConnectionPluginManager> instances = new CopyOnWriteArrayList<>();
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java:52]: https://stackoverflow.com/questions/39573764/merging-list-concurrently-which-is-better-copyonwritearraylist-or-concurrentl
I suspect there will be way more writes than reads to this list so I propose we use ConcurrentLinkedQueue instead, what do you think?

File: src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java (hunk 1 above)
Comment (ColinKYuen) [reply src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java]: Test removed.

File: src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java (hunk 2 above)
Comment (ColinKYuen) [reply src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java:57]: Test removed.

File: src/main/user-impl/java/com/mysql/cj/jdbc/AbandonedConnectionCleanupThread.java (hunk 3 above)
Comment (ColinKYuen) [reply src/main/user-impl/java/com/mysql/cj/jdbc/AbandonedConnectionCleanupThread.java:54]: Removed from class.

File: src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java (hunk 4 above)
Comment (ColinKYuen) [reply src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java:20]: Test removed.

File: src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java (hunk 5 above)
Comment (ColinKYuen) [reply src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java:41]: Test removed.

File: src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java (hunk 6 above)
Comment (ColinKYuen) [reply src/test/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManagerTest.java:47]: Test removed.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java (hunk 7 above)
Comment (ColinKYuen) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java:52]: Sounds good. Will change this.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java [hunk 8]
```diff
@@ -47,18 +49,23 @@
   protected PropertySet propertySet = null;
   protected IConnectionPlugin headPlugin = null;
   ClusterAwareConnectionProxy proxy;
-  private final Thread shutdownHook = new Thread(this::releaseResources);
+  protected static final Queue<ConnectionPluginManager> instances;
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java:52]:
```suggestion
  protected static final Queue<ConnectionPluginManager> instances = new ConcurrentLinkedQueue<>(); 
```

File: src/main/user-impl/java/software/aws/rds/jdbc/mysql/Driver.java [hunk 9]
```diff
@@ -86,4 +87,11 @@ public static void setAcceptAwsProtocolOnly(boolean awsProtocolOnly) {
   public Driver() throws SQLException {
     // Required for Class.forName().newInstance()
   }
+
+  /**
+   * Release all resources currently held up by Connection Plugin Manager
+   */
+  public static void releasePluginManager() {
```
Comment (karenc-bq) [inline src/main/user-impl/java/software/aws/rds/jdbc/mysql/Driver.java:94]:
```suggestion
  public static void releasePluginManagers() {
```

File: src/main/user-impl/java/software/aws/rds/jdbc/mysql/Driver.java [hunk 10]
```diff
@@ -86,4 +87,11 @@ public static void setAcceptAwsProtocolOnly(boolean awsProtocolOnly) {
   public Driver() throws SQLException {
     // Required for Class.forName().newInstance()
   }
+
+  /**
+   * Release all resources currently held up by Connection Plugin Manager
```
Comment (karenc-bq) [inline src/main/user-impl/java/software/aws/rds/jdbc/mysql/Driver.java:92]:
```suggestion
   * Release all resources currently held up by {@link ConnectionPluginManager}.
```

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java [hunk 11]
```diff
@@ -47,18 +49,23 @@
...
+
+  static {
+    instances = new ConcurrentLinkedQueue<>();
+    Runtime.getRuntime().addShutdownHook(new Thread(ConnectionPluginManager::releaseAllResources));
+  }
 
   public ConnectionPluginManager(Log logger) {
     if (logger == null) {
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java:60]: To be honest I am not sure if we want to throw a RuntimeException if a logger is not provided. What do you think of setting logger to a null logger instead @sergiyv-bitquill @ColinKYuen

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java (hunk 8 above)
Comment (ColinKYuen) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java:52]: Done.

File: src/main/user-impl/java/software/aws/rds/jdbc/mysql/Driver.java (hunk 9 above)
Comment (ColinKYuen) [reply src/main/user-impl/java/software/aws/rds/jdbc/mysql/Driver.java:94]: Done.

File: src/main/user-impl/java/software/aws/rds/jdbc/mysql/Driver.java (hunk 10 above)
Comment (ColinKYuen) [reply src/main/user-impl/java/software/aws/rds/jdbc/mysql/Driver.java:92]: Done.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java (hunk 11 above)
Comment (sergiyv-bitquill) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java:59]: I'd be against setting logger to null. That add complexity to the code like checking if logger is null or not. It's better to make sure to pass a dummy logger.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java [hunk 12]
```diff
@@ -47,18 +49,22 @@
   protected PropertySet propertySet = null;
   protected IConnectionPlugin headPlugin = null;
   ClusterAwareConnectionProxy proxy;
-  private final Thread shutdownHook = new Thread(this::releaseResources);
+  protected static final Queue<ConnectionPluginManager> instances = new ConcurrentLinkedQueue<>();
```
Comment (sergiyv-bitquill) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java:52]: Nit: Static variables are usually defined at the top of class declaration, before other variables.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/ClusterAwareConnectionProxy.java [hunk 13]
```diff
@@ -57,11 +57,11 @@
 import com.mysql.cj.util.StringUtils;
 import com.mysql.cj.util.Util;
 
+import javax.net.ssl.SSLException;
```
Comment (sergiyv-bitquill) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/ClusterAwareConnectionProxy.java:60]: Do we need this package?

Comment (ColinKYuen) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/ClusterAwareConnectionProxy.java:60]: This was already part of the original code, I just used the IDE's optimize import feature and it moved it up.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java (hunk 12 above)
Comment (ColinKYuen) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java:52]: Done.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java (hunk 11 above)
Comment (karenc-bq) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/ConnectionPluginManager.java:59]: I was referring to the class `com.mysql.cj.log.NullLogger`. I think we are on the same page here.

**Review by karenc-bq — APPROVED** [review]

**Review by sergiyv-bitquill — APPROVED** [review]

**Pull request merged**